nemo/collections/asr/parts/submodules/multitask_decoding.py

```diff
@@ -0,0 +1,492 @@
+
+            # extract the hypotheses
+            hypotheses_list = hypotheses_list[0]  # type: List[Hypothesis]
+
+        prediction_list = hypotheses_list
+
+        if isinstance(prediction_list[0], NBestHypotheses):
+            hypotheses = []
+            all_hypotheses = []
+
+            for nbest_hyp in prediction_list:  # type: NBestHypotheses
+                n_hyps = nbest_hyp.n_best_hypotheses  # Extract all hypotheses for this sample
+                decoded_hyps = self.decode_hypothesis(n_hyps)
+
+                hypotheses.append(decoded_hyps[0])  # best hypothesis
+                all_hypotheses.append(decoded_hyps)
+
+            if return_hypotheses:
+                return hypotheses, all_hypotheses
+
+            best_hyp_text = [h.text for h in hypotheses]
+            all_hyp_text = [h.text for hh in all_hypotheses for h in hh]
+            return best_hyp_text, all_hyp_text
+
+        else:
+            hypotheses = self.decode_hypothesis(prediction_list)
+
+            if return_hypotheses:
+                return hypotheses, None
+
+            best_hyp_text = [h.text for h in hypotheses]
+            return best_hyp_text, None
+
+    def decode_hypothesis(self, hypotheses_list: List[Hypothesis]) -> List[Union[Hypothesis, NBestHypotheses]]:
+        """
+        Decode a list of hypotheses into a list of strings.
+
+        Args:
+            hypotheses_list: List of Hypothesis.
+
+        Returns:
+            A list of strings.
+        """
+        for ind in range(len(hypotheses_list)):
+            # Extract the integer encoded hypothesis
+            prediction = hypotheses_list[ind].y_sequence
+
+            if type(prediction) != list:
+                prediction = prediction.tolist()
+
+            hypothesis = self.decode_tokens_to_str(prediction)
+
+            if self.compute_hypothesis_token_set:
+                hypotheses_list[ind].tokens = self.decode_ids_to_tokens(prediction)
+
+            # De-tokenize the integer tokens
+            hypotheses_list[ind].text = hypothesis
+
+        return hypotheses_list
+
+    @abstractmethod
+    def decode_tokens_to_str(self, tokens: List[int]) -> str:
+        """
+        Implemented by subclass in order to decoder a token id list into a string.
+
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A decoded string.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
+        """
+        Implemented by subclass in order to decode a token id list into a token list.
+        A token list is the string representation of each token id.
+
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A list of decoded tokens.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def decode_tokens_to_lang(self, tokens: List[int]) -> str:
```

Comment:
Looks specific to aggregate tokenizer. should we include this in abstract class.

Reply:
I agree, but RNNT has similar API for some reason. Well it would be necessary for other tokenizers to also use this function for language calculation, and it only gets called if used explicitly so it might be ok for now. But i do agree this is too specific to be part of abstract class. Fyi @bmwshop @KunalDhawan